adminer/drivers/clickhouse.inc.php

```diff
@@ -0,0 +1,389 @@
+            }
+        }
+        return $return;
+    }
+
+    function is_view($table_status) {
+        return false;
+    }
+
+    function fk_support($table_status) {
+        return false;
+    }
+
+    function convert_field($field) {
+
+    }
+
+    function unconvert_field($field, $return) {
+        if (in_array($field['type'], ["Int8", "Int16", "Int32", "Int64", "UInt8", "UInt16", "UInt32", "UInt64", "Float32", "Float64"])) {
+            return "to".$field['type']."(".$return.")";
+        }
+
+        return $return;
+    }
+
+    function fields($table) {
+        $return = array();
+        $result = get_rows("SELECT name, type, default_expression FROM system.columns WHERE ".idf_escape('table')." = ".q($table));
+
+        foreach($result as $row) {
+            $type = trim($row['type']);
+
+            $nullable = strpos($type, 'Nullable(') === 0;
+
+            $return[trim($row['name'])] = array(
+                "field" => trim($row['name']),
+                "full_type" => $type,
+                "type" => $type,
+                "default" => trim($row['default_expression']),
+                "null" => $nullable,
+                "auto_increment" => '0',
+                "privileges" => array("insert" => 1, "select" => 1, "update" => 0),
+            );
+        }
+
+        return $return;
+    }
+
+    function indexes($table, $connection2 = null) {
+        return array();
+    }
+
+    function foreign_keys($table) {
+        return array();
+    }
+
+    function collations() {
+        return array();
+    }
+
+    function information_schema($db) {
+        return false;
+    }
+
+    function error() {
+        global $connection;
+        return h($connection->error);
+    }
+
+    function types() {
+        return array();
+    }
+
+    function schemas() {
+        return array();
+    }
+
+    function get_schema() {
+        return "";
+    }
+
+    function set_schema($schema) {
+        return true;
+    }
+
+    function auto_increment()
+    {
+        return '';
+    }
+
+    function support($feature) {
+        return preg_match("~^(columns|sql|status|table)$~", $feature);
+    }
+
+    $jush = "clickhouse";
+    $types = array();
+    $structured_types = array();
+    foreach (array( //! arrays
+                 lang('Numbers') => array("Int8" => 3, "Int16" => 5, "Int32" => 10, "Int64" => 19, "UInt8" => 3, "UInt16" => 5, "UInt32" => 10, "UInt64" => 20, "Float32" => 7, "Float64" => 16),
+                 lang('Date and time') => array("Date" => 13, "DateTime" => 20),
+                 lang('Strings') => array("String" => 0),
+                 lang('Binary') => array("FixedString" => 0),
+             ) as $key => $val) {
+        $types += $val;
+        $structured_types[$key] = array_keys($val);
+    }
+    $unsigned = array();
+    $operators = array("=", "<", ">", "<=", ">=", "!=", "~", "!~", "LIKE", "LIKE %%", "IN", "IS NULL", "NOT LIKE", "NOT IN", "IS NOT NULL"); // no "SQL" to avoid SQL injection
```

Comment:
Do `SELECT` queries allow data modification in ClickHouse?

Reply:
I am not sure, but looks like it is not possible. Can I add "SQL" operator?

Reply:
Yes.